Extensions: Clarify documentation. (#231)
pals-project/pals: source/extensions.md

> that has no analogue with other machines or simulation programs can be put in as an extension. 

Reviewer: This is duplicated from definitions.md, which is not ideal, because the two texts will likely end up being out of sync. That might be already the case, given that the text in definitions.md reads
```
An Extension is data added to a PALS lattice that is not part of the PALS standard.
For example, information specific to a particular machine that has no analogue with other
machines can be put in as an extension. A list of "registered" extensions is in the 
[Extension](#c:extension) section.
```
hence uses `PALS lattice` instead of `PALS file` and mentions only `machine` instead of `machine and simulation program`. 

I would recommend having a one-liner in definitions.md and leaving the verbose description only in extensions.md.

Author: I changed `PALS lattice` to `PALS file` and added `machine or simulation program`. There are only two sentences in definitions.md and I don't mind if they get out of synch as long as they are correct.

Maintenance of an extension schema falls outside of PALS governance. 

> Extension maintainers can add extension documentation submitting a [pull request](#s:contribute).

Reviewer: ```suggestion
Extension maintainers can add extension documentation by submitting a [pull request](#s:contribute).
```

If the amount of documentation is "large", It is encouraged that the documentation put in the PALS

> repo should be relatively brief and there should be a reference to full documentation maintained

Reviewer: ```suggestion
If the amount of documentation is "large", it is encouraged that the documentation put in the PALS
repository should be relatively brief and there should be a reference to full documentation maintained
```

Maintenance of an extension schema falls outside of PALS governance. 
Extension maintainers can add extension documentation by submitting a [pull request](#s:contribute).
If the amount of documentation is "large", the documentation put in the PALS
repository should be relatively brief and there should be a reference to full documentation maintained
elsewhere, if such documentation exists.
The advantage of this setup is that since the full documentation is external to the PALS repo,
extension maintainers are not encumbered to use PALS pull requests when modifying the full documentation.

Syntax for including extension data in a PALS file is in the 
[Extension Syntax](#s:extension-syntax) section.
A list of documented extensions is in the [Extension Documentation](#s:extension-list) section
and this is where extension maintainers should put their extension documentation.

%---------------------------------------------------------------------------------------------------
(s:extension-syntax)=
## Extension Syntax

An extension block can be introduced at any level in a PALS file.
The syntax is:
```{code} yaml
<some_name>:
  extension: <extension-type>
  ... extension info ...
```

> where <some_name> is an arbitrary key whose value is a dictionary,

Reviewer: Backticks were dropped:

```suggestion
where `<some_name>` is an arbitrary key whose value is a dictionary,
```

```
In this example, the `synch_connect` dictionary key is the start of the extension. The type of extension is
`Cornell_CESR_Connect`. Any further entries in `synch_connect` will be excluded from any
PALS validation process. Either a dict or a list can be used for the children of `<some_name>`.

Alternatively, extension metadata names may be "registered" using an `extension_labels` attribute.
Registration means that a PALS parser can validate the extension name.
Extension metadata names appear at the start (top level) of an extension block.
The `extension_labels` attribute must appear as a child of the `PALS` root node.

> The `extension_labels` block can also be used to define enum values.

Reviewer: We don't explain how an "enum" is defined versus an ordinary name. From the example, an enum value is just a `names` entry that happens to be matched as a value rather than a key (is it?). I think a sentence making that explicit could help, since  we now state twice that "enums can be defined" but never really show it as a distinct mechanism.

Author: I added a definition for `enum` in definitions.md and added a link to the def in extensions.md.

> The `extension_labels` block has three components, at least one of them is required:

Reviewer: Here we use both "`extension_labels` block" and "`extension_labels` attribute". Could we use only one wording (either "block" or "attribute") consistently instead?

Author: Changing to `block`.

PALS:
  extension_labels:
    names:
      data_base:   Name used in the machine database.
      blueprint:   Blueprint ID number. 
      Rotator:     Enum value. A new type of element.
    prefixes:
      SciBmad_:    Prefix for the SciBmad ecosystem.

  facility:
    - Q1w:                   # Lattice element name
        data_base: Q1_czt    # Extension matching "data_base".
        SciBmad_connect:     # Extension with "SciBmad_" prefix.
          ... SciBmad connection info ...
    ...
    - R23:              # Another lattice element
        kind: Rotator   # enum value is a documented extension.
```

> The advantage of using `extension_labels` is that enums can be defined and 
> single attribute key extensions are possible (there does not have to be an `extension` key).

Reviewer 2: This might be a bit too hard to read and understand.
I would suggest

```suggestion
The advantage of using `extension_labels` is that is can be used to mask attributes at any level.
```

Also the example above does not show an enum...

Author: The phrasing  `mask attributes` is cryptic to me. Attributes are not being masked (hidden).

Reviewer 3: @DavidSagan I agree. "mask attributes" is not clear to me.
